Add hcfoldMap, hctraverse etc.

phadej opened the pull request:
This would complete Applicative-like story. 

Maybe it worth adding `HTraverse` (or method to `HSequence` ?)

--

The primary motivation is `hcfoldMap` (which you *cannot* implement generically in terms of `hcollapse`!)

---

@kosmikus are you ok if we add these functions?

phadej commented:
Questions:
- do we need `htraverse` (non-constrained). Yes for completeness, probably no in practice.
- do we need `hcfoldMap` in a class? (same with `hsequence'`) Should they be derived?

kosmikus commented on src/Generics/SOP/Classes.hs:
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
+  -- /Instances:/
+  --
+  -- @
+  -- ...
```
These should be added.

kosmikus commented on src/Generics/SOP/Classes.hs:
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
+  -- | Corresponds to 'Data.Foldable.traverse_'.
+  hctraverse_ :: (AllN h c xs, Applicative g) => proxy c -> (forall a. c a => f a -> g ()) -> h f xs -> g ()
+
+-- | Flipped version of 'hcfor_'
```
Should this read "of `hctraverse_`"?

kosmikus commented on src/Generics/SOP/Classes.hs:
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
+  hcfoldMap p f = unK . hctraverse_ p (K . f)
+
+  -- | Corresponds to 'Data.Foldable.traverse_'.
+  hctraverse_ :: (AllN h c xs, Applicative g) => proxy c -> (forall a. c a => f a -> g ()) -> h f xs -> g ()
```
What about `hcfoldr`? Should a connection be established to the catamorphisms?

kosmikus commented:
Regarding your other questions:

- Yes, I'd always add the unconstrained versions of the functions for completeness.
- I think the assumption of `generics-sop` so far is that the only types we're interested in are `NP`, `POP`, `NS` and `SOP`, and so all the `H*` classes have at most four instances. As long as we wouldn't want a deviating definition for one of these, it's completely ok to define functions outside of the classes.

kosmikus commented:
Oh, and could you please tell me once again why `hcfoldMap` is not implementable in terms of `hcollapse`?

phadej commented:
hcollapse :: ... -> k, [k] or [[k]] depending on the structure (NS, NP and SOP, POP respectively). hcfoldMap is always -> m. From my limited pov, hctraverse_ is more useful variant, maybe because monadic DSLs are more common. But hcfoldMap ja hctraverse_ can be defined in terms of each other.

Sent from my iPhone

> On 1 Nov 2017, at 10.30, Andres Löh <notifications@github.com> wrote:
> 
> Oh, and could you please tell me once again why hcfoldMap is not implementable in terms of hcollapse?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

phadej commented:
@kosmikus I added `htraverse_` and `htraverse'` as derived functions (`hpure` is class member), so there is `AllN h Top cs` - trivial constraint. Or is it better to have these functions as members?

Otherwise I think I addressed all your comments & concerns.

kosmikus commented:
@phadej Defining the non-constrained variants in terms of the constrained variants by instantiating the constraint to `Top` is currently not idiomatic, i.e., not consistent with the rest of the package. These should instead just have `SListIN` constraints.

I do want to utlimately reorganise things such that `SListI ~ All Top`, but last time I tried, this requires abandoning pre-ghc-8 compatibility for certain technical reasons (IIRC, something to do with undecidable superclasses).

kosmikus commented:
@phadej I have now merged this, but also made the change to implement the unconstrained variants directly. I also cleaned up a bit (comments, exports). There's still the question whether more of the unconstrained variants or specialized versions should be defined, but it's of somewhat lesser importance.

phadej commented:
@kosmikus great. Things are easier to add, than to remove. So let's see if we'll need them :)
